lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -2080,6 +2086,38 @@ static LogicalResult checkConnectFlow(Operation *connect) {
     return diag.attachNote(dstRef.getLoc())
            << "the destination was defined here";
   }
+
+  // If the connect is for RefType, implement the constraint for downward only
+  // references. We cannot connect
+  //  1. an input reference port to the output reference port.
+  //  2. instance reference results to each other.
+  //  This means, the connect can only be used for forwarding RefType module
+  //  ports to Instance ports.
+  if (dst.getType().isa<RefType>()) {
+    if (getDeclarationKind(src) == getDeclarationKind(dst)) {
+      auto srcRef = getFieldRefFromValue(src);
+      bool rootKnown;
+      auto srcName = getFieldName(srcRef, rootKnown);
+      auto diag = emitError(connect->getLoc());
+      diag << "connect has invalid flow for Ref type ports: the source "
+              "expression ";
+      if (rootKnown)
+        diag << "\"" << srcName << "\" ";
```

Comment:
This should do something if the root is not known other than print a confusing error message. Ditto below.